### DISCUSSION TITLE
rtsp stream transcode h265 to vp8 stream over kvswebrtc

### DISCUSSION THREAD

**Discussion by pmuvvagithub** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#1772] [Q&A] [answered]
gst-launch-1.0 rtspsrc location=rtsp://3.84.79.165:31501/burglary ! application/x-rtp,encoding-name=H265 ! rtpjitterbuffer ! rtph265depay ! h265parse ! video/x-h265, alignment=au ! avdec_h265 ! vp8enc ! kvswebrtc channel-name="fc59c012-dc76-40eb-a550-54b06678c92d"

When I am trying to trancode h265 video to vp8 I am getting an error: 
erroneous pipeline: could not link vp8enc0 to kvsplugin0

Please let me know how to solve this problem?

**Comment by pmuvvagithub** [comment]
it works fine when I transcode from h265 to h264:
gst-launch-1.0 rtspsrc location=rtsp://3.84.79.165:31501/burglary ! application/x-rtp,encoding-name=H265 ! rtpjitterbuffer ! rtph265depay ! h265parse ! video/x-h265, alignment=au ! avdec_h265 ! x264enc ! video/x-h264, alignment=au ! kvswebrtc channel-name="fc59c012-dc76-40eb-a550-54b06678c92d" 

But it fails for vp8.

**Comment by disa6302** [comment]
Hello,

We do not provide a webrtc gstreamer plugin. Can you provide more details on what the kvswebrtc plugin is?

**Comment by pmuvvagithub** [comment]
It is taken from amazon-kinesis-video-streams-webrtc-sdk-c gstream sample plugin

**Comment by disa6302** [comment]
@pmuvvagithub , our gstreamer samples do not have a kvswebrtc. Please point to the code you are using. We would be unable to help without the required details.

Also, the issue is potential not related to KVS but the issue is with the source pipeline. Please try your source pipeline out without using kvs, you could try filesink or fakesink to ensure your pipeline is contructed properly before incorporating kvs.

**Comment by hassanctech** [comment]
@pmuvvagithub 
> When I am trying to trancode h265 video to vp8 I am getting an error:
erroneous pipeline: could not link vp8enc0 to kvsplugin0

Does your custom GStreamer plugin advertise support for vp8 in its source pad?  For the elements to link the sink pad of vp8enc needs to agree with the src pad of the sink plugin.

Once that is resolved, you need to make sure in your implementation if you're running as master you create a local transceiver advertising VP8 support so when an offer is received (remote transceiver) we can respond with the correct answer if VP8 is present in a local transceiver.

**Comment by hassanctech** [comment, marked as answer]
@pmuvvagithub Is this what you're using:  https://github.com/aws-samples/amazon-kinesis-video-streams-demos/tree/master/gst/gst-kvs-plugin ?

I should clarify here this is in our samples/demos repo and not something that has been actively maintained or thoroughly tested.  If this is indeed what you're using then this is where you would specify the support for any additional codecs:  https://github.com/aws-samples/amazon-kinesis-video-streams-demos/blob/master/gst/gst-kvs-plugin/src/GstPlugin.c#L104-L105

Finally I would really recommend you spend some time to get familiar with how GStreamer works it will help you a lot in debugging issues such as this question is really a question about the fundamentals of GStreamer and completely unrelated to KVS.  Once you're familiar with you source / sink and elements behave and what the pads do and how they need to be set up to "agree" so they can be linked I think you will save yourself a lot of time and headache instead of just trying to hook up components randomly and then not understanding why they cannot link.  I hope this helps!
